File: app/Sudoku Battles/Data/FirestoreDs.swift
import FirebaseFirestore
import FirebaseAuth

class FirestoreDs {
    static let shared = FirestoreDs()
    static let usersCollectionId = "users"
    static let matchmakingCollectionId = "matchmaking"
    static let playerDuelsCollectionId = "player-duels"
    static let botDuelsCollectionId = "bot-duels"
    
    private let firestore: Firestore
    private let matchmaking: CollectionReference
    private let users: CollectionReference

    init() {
        self.firestore = Firestore.firestore()
        if ProcessInfo.dev {
            let settings = Firestore.firestore().settings
            settings.host = "\(ProcessInfo.firebaseHost):\(ProcessInfo.firestorePort)"
            settings.isSSLEnabled = false
            settings.cacheSettings = MemoryCacheSettings()
            Firestore.firestore().settings = settings
        }
                
        self.matchmaking = firestore.collection(Self.matchmakingCollectionId)
        self.users = firestore.collection(Self.usersCollectionId)
    }
    
    func reference(of path: String) -> DocumentReference {
        return firestore.document(path)
    }
    
    func updateDocument(_ reference: DocumentReference, fields: [String: Any]) async throws {
        try await reference.updateData(fields)
    }
    func updateDocument(_ path: String, fields: [String: Any]) async throws {
        try await updateDocument(firestore.document(path), fields: fields)
    }

    func deleteDocument(_ reference: DocumentReference) async throws {
        try await reference.delete()
    }
    func deleteDocument(_ path: String) async throws {
        try await deleteDocument(firestore.document(path))
    }


    func getDocument<T : Decodable>(_ reference: DocumentReference) async throws -> T? {
        var document: DocumentSnapshot!
        do {
            document = try await reference.getDocument()
            if !document.exists { return nil }
        } catch {
            logger.error("\(error)")
            throw AppError.networkError
        }
        
        do {
            return try document.data(as: T.self)
        }
        catch {
            logger.error("\(error)")
            throw AppError.invalidResponse
        }
    }
    func getDocument<T : Decodable>(_ path: String) async throws -> T? {
        return try await getDocument(firestore.document(path))
    }

    func subscribeToDocument<T : Decodable>(_ reference: DocumentReference, callback: @escaping (T) -> ()) async throws -> ListenerRegistration {
        let document = try? await reference.getDocument()
        if let data = try? document?.data(as: T.self) { callback(data) }
        return reference.addSnapshotListener { (documentSnapshot, error) in
            guard let document = documentSnapshot,
                  document.exists,
                  let data = try? document.data(as: T.self),
                  error == nil
            else { return }

            callback(data)
        }
    }
    func subscribeToDocument<T : Decodable>(_ path: String, callback: @escaping (T) -> ()) async throws -> ListenerRegistration {
        return try await subscribeToDocument(firestore.document(path), callback: callback)
    }

        
    func queryUserData(uid: String) async throws -> UserData? {
        return try await getDocument(users.document(uid))
    }
    func queryUserDatas(usernamePartial: String) async throws -> [String : UserData]? {
        var query: QuerySnapshot!
        do {
            query = try await users
                .whereField("usernameLowercase", isGreaterThanOrEqualTo: usernamePartial.lowercased())
                .whereField("usernameLowercase", isLessThanOrEqualTo: usernamePartial.lowercased() + "\u{f8ff}")
                .limit(to: 5)
                .getDocuments()
        } catch {
            logger.error("\(error)")
            throw AppError.networkError
        }

        var results: [String: UserData] = [:]
        for document in query.documents {
            if let userData = try? document.data(as: UserData.self) {
                results[document.documentID] = userData
            }
        }
        return results
    }
    func updateFcmToken(uid: String, fcmToken: String, deviceId: UUID) async throws {
        try await updateDocument(users.document(uid), fields: ["fcmTokens.\(deviceId.uuidString.lowercased())" : fcmToken])
    }
    
    func deleteMatchmaking(uid: String) async throws {
        try await deleteDocument(matchmaking.document(uid))
    }
    
}

File: app/web/src/data/FirestoreDs.ts
import {
    getFirestore,
    collection,
    doc,
    getDoc,
    updateDoc,
    deleteDoc,
    query,
    where,
    limit,
    getDocs,
    Firestore,
    CollectionReference,
    DocumentReference,
    QuerySnapshot,
    onSnapshot,
} from "firebase/firestore";
import firebaseApp from "../utility/FirebaseApp";
import { AppError } from "./Model/AppError";
import type UserData from "./Model/UserData";

export default class FirestoreDs {
    private static firestore: Firestore = getFirestore(firebaseApp());

    private static usersCollection: CollectionReference = collection(
        FirestoreDs.firestore,
        "users"
    );
    private static matchmakingCollection: CollectionReference = collection(
        FirestoreDs.firestore,
        "matchmaking"
    );

    private static getDocumentReference(path: string): DocumentReference {
        return doc(FirestoreDs.firestore, path);
    }

    static async updateDocument(
        reference: DocumentReference,
        fields: Record<string, any>
    ): Promise<void> {
        try {
            await updateDoc(reference, fields);
        } catch (error) {
            console.error(error);
            throw AppError.NetworkError;
        }
    }

    static async deleteDocument(reference: DocumentReference): Promise<void> {
        try {
            await deleteDoc(reference);
        } catch (error) {
            console.error(error);
            throw AppError.NetworkError;
        }
    }

    static async getDocument<T>(
        reference: DocumentReference
    ): Promise<T | null> {
        try {
            const document = await getDoc(reference);
            if (document.exists()) {
                return document.data() as T;
            }
            return null;
        } catch (error) {
            console.error(error);
            throw AppError.InvalidResponse;
        }
    }

    static async subscribeToDocument<T>(
        reference: DocumentReference,
        callback: (data: T) => void
    ): Promise<void> {
        try {
            onSnapshot(reference, (snapshot) => {
                if (snapshot.exists()) {
                    callback(snapshot.data() as T);
                }
            });
        } catch (error) {
            console.error(error);
            throw AppError.NetworkError;
        }
    }

    static async queryUserData(uid: string): Promise<UserData | null> {
        const userRef = doc(FirestoreDs.usersCollection, uid);
        return this.getDocument(userRef);
    }

    static async queryUserDatas(
        usernamePartial: string
    ): Promise<Record<string, UserData>> {
        try {
            const q = query(
                FirestoreDs.usersCollection,
                where("usernameLowercase", ">=", usernamePartial.toLowerCase()),
                where(
                    "usernameLowercase",
                    "<=",
                    usernamePartial.toLowerCase() + "\uf8ff"
                ),
                limit(5)
            );
            const querySnapshot: QuerySnapshot = await getDocs(q);

            const results: Record<string, UserData> = {};
            querySnapshot.forEach((doc) => {
                results[doc.id] = doc.data() as UserData;
            });

            return results;
        } catch (error) {
            console.error(error);
            throw AppError.NetworkError;
        }
    }

    static async queryUserDatasByUids(
        uids: string[]
    ): Promise<Record<string, UserData>> {
        try {
            const q = query(
                FirestoreDs.usersCollection,
                where("__name__", "in", uids),
                limit(5)
            );
            const querySnapshot: QuerySnapshot = await getDocs(q);

            const results: Record<string, UserData> = {};
            querySnapshot.forEach((doc) => {
                results[doc.id] = doc.data() as UserData;
            });

            return results;
        } catch (error) {
            console.error(error);
            throw AppError.NetworkError;
        }
    }

    static async updateFcmToken(
        uid: string,
        fcmToken: string,
        deviceId: string
    ): Promise<void> {
        try {
            const userRef = doc(FirestoreDs.usersCollection, uid);
            const fields = {
                [`fcmTokens.${deviceId.toLowerCase()}`]: fcmToken,
            };
            await this.updateDocument(userRef, fields);
        } catch (error) {
            console.error(error);
            throw AppError.NetworkError;
        }
    }

    static async deleteMatchmaking(uid: string): Promise<void> {
        try {
            const matchmakingRef = doc(FirestoreDs.matchmakingCollection, uid);
            await this.deleteDocument(matchmakingRef);
        } catch (error) {
            console.error(error);
            throw AppError.NetworkError;
        }
    }
}
